### PR TITLE
ipn/ipnlocal,net/dns/resolver: serve MagicDNS names from live indexes

### PR DESCRIPTION
**Pull request by bradfitz**
Every netmap change, including an incremental delta of a single peer,
rebuilt the full MagicDNS state twice: dnsConfigForNetmap walked all
peers to build the dns.Config.Hosts map, and resolver.SetConfig then
walked that map again to build its reverse (PTR) index. On a tailnet
with 10k peers that is a lot of garbage per delta.

Instead, add a resolver.MagicDNSHosts hook, installed once by
LocalBackend, that the quad-100 resolver consults on demand at query
time. It is backed by nodeBackend's nodeByName, nodeByAddr, and peers
indexes, which are already maintained incrementally as netmap deltas
arrive. The subdomain-resolve capability check also moves to the hook
(checking the node's CapMap at query time), so dns.Config's
SubdomainHosts is no longer populated.

dns.Config.Hosts remains for control's DNS.ExtraRecords, which are
few and which feed the split-DNS decisions in dns.Manager's
compileConfig, and on Windows it still carries every node's records
because the hosts-file fallback path (compileHostEntries) needs the
complete enumerable set. Those compileConfig decisions also consulted
the per-node Hosts entries (hasHostsWithoutSplitDNSRoutes), so a new
Config.MagicDNSHostsUnrouted bit preserves that signal now that node
records are not listed: with MagicDNS names present but MagicDNS
domain routing off, quad-100 stays in the OS resolver path.

One small behavior change: reverse (PTR) lookups now also answer for
node addresses whose forward records are filtered out by the
IPv6-suppression rule (issue #1152), since nodeByAddr indexes all node
addresses. Previously such addresses were absent from the pushed
Hosts map and thus from the reverse index.

Updates #12542
Updates tailscale/corp#43949

### CODE REVIEW

File: ipn/ipnlocal/node_backend.go [hunk 1]
```diff
@@ -966,10 +972,119 @@ func (nb *nodeBackend) setFilter(f *filter.Filter) {
...
+	if nm := nb.netMap; nm != nil && nm.SelfNode.Valid() && nm.SelfNode.ID() == n.ID() {
+		return nm.AllCaps.Contains(tailcfg.NodeAttrDNSSubdomainResolve)
+	}
+	return n.CapMap().Contains(tailcfg.NodeAttrDNSSubdomainResolve)
+}
+
+// nodeByFQDNLocked returns the node (peer or self) with the given
+// MagicDNS FQDN. nb.mu must be held.
+func (nb *nodeBackend) nodeByFQDNLocked(fqdn dnsname.FQDN) (_ tailcfg.NodeView, ok bool) {
+	nid, ok := nb.nodeByName[strings.TrimSuffix(string(fqdn), ".")]
```
Comment (nickkhyl) [inline ipn/ipnlocal/node_backend.go]: Should we convert `fqdn` to lowercase before looking it up?

File: net/dns/resolver/tsdns.go [hunk 2]
```diff
@@ -229,6 +229,49 @@ type Resolver struct {
...
+// DNS.ExtraRecords entries, which are few, and is consulted first.
+//
+// Implementations must be safe for concurrent use and cheap: the
+// methods are called on the DNS query serving path.
+type MagicDNSHosts interface {
+	// LookupHost returns the IPs to answer for the node with the
+	// given MagicDNS FQDN, and whether the name is known. A known
+	// name may have no IPs of the queried family; that is "name
+	// exists, no records", not NXDOMAIN.
+	LookupHost(dnsname.FQDN) (ips []netip.Addr, ok bool)
```
Comment (nickkhyl) [inline net/dns/resolver/tsdns.go:256]: The docstring mentions "the queried family", but `LookupHost` doesn't really allow to query addresses of a given family.

IIUC, `LookupHost` returns all supported/enabled IP addresses (depending on `NodeAttrMagicDNSPeerAAAA`, whether self node has IPv4 addresses, etc.), and it's the caller's responsibility to filter them based on the queried family.

File: ipn/ipnlocal/node_backend.go [hunk 3]
```diff
@@ -1203,6 +1284,23 @@ func dnsConfigForNetmap(nm *netmap.NetworkMap, peers map[tailcfg.NodeID]tailcfg.
...
+		haveNodeRecords := nm.SelfName() != "" && nm.GetAddresses().Len() > 0
+		if !haveNodeRecords {
+			for _, peer := range peers {
+				if peer.Name() != "" && peer.Addresses().Len() > 0 {
+					haveNodeRecords = true
+					break
+				}
+			}
+		}
+		dcfg.MagicDNSHostsUnrouted = haveNodeRecords
```
Comment (nickkhyl) [inline ipn/ipnlocal/node_backend.go]: The logic here and in `(net/dns.Config).hasHostsWithoutSplitDNSRoutes` seems to match what we had prior to these changes when we used to set `dcfg.Hosts` on all platforms rather than just on Windows.

But can you think of a scenario where `haveNodeRecords`, and by extension `dcfg.MagicDNSHostsUnrouted`, would be `false` on a tailnet with MagicDNS disabled?

Wouldn't `haveNodeRecords := nm.SelfName() != "" && nm.GetAddresses().Len() > 0` always evaluate to `true` if we have a netmap?

File: ipn/ipnlocal/state_test.go [hunk 4]
```diff
@@ -1572,6 +1559,7 @@ func TestEngineReconfigOnStateChange(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			lb, engine, cc := newLocalBackendWithMockEngineAndControl(t, enableLogging)
+			lb.goos = "linux" // so the expectations below are the same on every host OS
```
Comment (nickkhyl) [inline ipn/ipnlocal/state_test.go:1562]: Since we still expect `dns.Config.Hosts` to be set on Windows, can you please either add a test for that specifically or re-add `hostsFor`, making it return an empty map on platforms other than Windows while keeping the old logic on Windows?

I don't love that this needs to work differently on Windows, but some of our customers depend on the `hosts` file, so http://go/beyonce on their behalf 😅

Comment (nickkhyl) [reply ipn/ipnlocal/state_test.go:1562]: Never mind. We kept a few Windows checks in `TestDNSConfigForNetmap`.

File: ipn/ipnlocal/node_backend.go (hunk 1 above)
Comment (bradfitz) [reply ipn/ipnlocal/node_backend.go]: Done, and documented on the interface method

File: net/dns/resolver/tsdns.go (hunk 2 above)
Comment (bradfitz) [reply net/dns/resolver/tsdns.go:256]: Reworded.

File: ipn/ipnlocal/node_backend.go (hunk 3 above)
Comment (bradfitz) [reply ipn/ipnlocal/node_backend.go]: simplified to just check for nm.SelfNode.Valid()

**Comment by bradfitz** [comment]
(Main diff before rebase atop main was https://github.com/tailscale/tailscale/compare/b85f0e73faa419321051045f3fd4e7cb7b582f09..e36dca574ad4de078144bc80c11f0ac31c38f556)

**Review by nickkhyl — APPROVED** [review]

File: ipn/ipnlocal/node_backend.go [hunk 5]
```diff
@@ -1126,10 +1132,122 @@ func (nb *nodeBackend) setFilter(f *filter.Filter) {
...
+	}
+	return nb.nodeByIDLocked(nid)
+}
+
+// magicDNSAddrs returns the MagicDNS answer addresses for a node
+// owning the given addresses: all of its IPv6 addresses if this node
+// (the querier) is IPv6-only, otherwise everything except the node's
+// IPv6 addresses when it also has IPv4, unless wantAAAA. The result
+// may be empty (name exists, no records of the desired family).
+func magicDNSAddrs(addrs views.Slice[netip.Prefix], selfV6Only, wantAAAA bool) (ips []netip.Addr) {
```
Comment (creachadair) [inline ipn/ipnlocal/node_backend.go:1223]: The two bools in a row thing seems a little risky (easy to flip them). Ordinarily since it's unexported I wouldn't worry too much about that, but almost all our tests are based on direct use of the unexported stuff.

**Review by creachadair — COMMENTED** [review]
(feel free to ignore this, I saw this merge while still reading)

File: ipn/ipnlocal/node_backend.go (hunk 5 above)
Comment (bradfitz) [reply ipn/ipnlocal/node_backend.go:1223]: Sent https://github.com/tailscale/tailscale/pull/20455
